src/main/scala/com/box/castle/core/committer/manager/BatchSizeManager.scala

```diff
@@ -0,0 +1,75 @@
+package com.box.castle.core.committer.manager
+
+import com.box.castle.core.committer.manager.BatchSizeManager._
+import com.box.castle.core.common.{BoundedQueue, ReadSample}
+import org.joda.time.Duration
+
+
+/**
+  * Implements logic to track reads and compute delay for following reads so as to ensure we always read close to
+  * bufferSize of data from Kafka.
+  *
+  */
+class BatchSizeManager(samplingSlots: Int, samplingInterval: Long, bufferSize: Int) {
+
+  private val samples = new BoundedQueue[ReadSample](samplingSlots)
+  private var consecutiveFullBuffers = 0
+
+  /**
+    * Tracks bytes read from kafka and the corresponding timestamp based on specified sampling rate
+    * Also keeps track of consecutive full buffer reads seen.
+    *
+    * @param bytesRead
+    * @param timestamp
+    */
+  def track(bytesRead: Int, timestamp: Long) = {
+    if (bytesRead >= bufferSize) {
+      // Update number of times we have seen consecutive full buffer reads
+      consecutiveFullBuffers += 1
+    }
+    else if (consecutiveFullBuffers > 0) {
+      // Reset counter if no more consecutive full buffer reads seen
+      consecutiveFullBuffers = 0
+    }
+
+    if (samples.isEmpty || timestamp - samples.last.timestamp > samplingInterval) {
+      // Add sample to the queue if sampling interval amount of time has passed else we discard it
+      samples.enqueue(ReadSample(bytesRead, timestamp))
```

Comment:
I am accumulating the total bytes read when I compute delay. But that needs a traversal of the queue so I guess this is a better option. Though I hope Long is sufficient for total data read . It can hold upto 9Exabyte which should be fine I guess.